File: dotnet_projects/OcrLiteOnnxVB/OcrLiteLib/DbNet.vb
Imports System.Drawing
Imports ClipperLib
Imports Emgu.CV
Imports Emgu.CV.CvEnum
Imports Emgu.CV.Structure
Imports Emgu.CV.Util
Imports Microsoft.ML.OnnxRuntime
Imports Microsoft.ML.OnnxRuntime.Tensors

Namespace OcrLiteLib
    Public Class DbNet
        Private ReadOnly MeanValues As Single() = {0.485F * 255.0F, 0.456F * 255.0F, 0.406F * 255.0F}
        Private ReadOnly NormValues As Single() = {1.0F / 0.229F / 255.0F, 1.0F / 0.224F / 255.0F, 1.0F / 0.225F / 255.0F}

        Private dbNet As InferenceSession

        Public Sub New(path As String, numThread As Integer)
            Try
                Dim op As New SessionOptions()
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED
                op.InterOpNumThreads = numThread
                op.IntraOpNumThreads = numThread
                dbNet = New InferenceSession(path, op)
            Catch ex As Exception
                Console.WriteLine(ex.Message + ex.StackTrace)
                Throw ex
            End Try
        End Sub

        Protected Overrides Sub Finalize()
            Try
                dbNet.Dispose()
            Finally
                MyBase.Finalize()
            End Try
        End Sub

        Public Function GetTextBoxes(src As Mat, scale As ScaleParam, boxScoreThresh As Single, boxThresh As Single, minArea As Single, unClipRatio As Single) As List(Of TextBox)
            Dim srcResize As New Mat()
            CvInvoke.Resize(src, srcResize, New Size(scale.DstWidth, scale.DstHeight))
            Dim inputTensors As Tensor(Of Single) = OcrUtils.SubstractMeanNormalize(srcResize, MeanValues, NormValues)
            Dim inputs As List(Of NamedOnnxValue) = New List(Of NamedOnnxValue) From {
                NamedOnnxValue.CreateFromTensor("input0", inputTensors)
            }

            Try
                Using results As IDisposableReadOnlyCollection(Of DisposableNamedOnnxValue) = dbNet.Run(inputs)
                    Dim resultsArray = results.ToArray()
                    Console.WriteLine(resultsArray)
                    Dim textBoxes = GetTextBoxes(resultsArray, srcResize.Rows, srcResize.Cols, scale, boxScoreThresh, boxThresh,
                        minArea, unClipRatio)
                    Return textBoxes
                End Using
            Catch ex As Exception
                Console.WriteLine(ex.Message + ex.StackTrace)
                Throw ex
            End Try
            Return Nothing
        End Function

        Private Shared Function GetTextBoxes(outputTensor As DisposableNamedOnnxValue(), rows As Integer, cols As Integer, s As ScaleParam, boxScoreThresh As Single, boxThresh As Single,
            minArea As Single, unClipRatio As Single) As List(Of TextBox)
            Dim rsBoxes As New List(Of TextBox)

            Dim outputData As Single() = outputTensor(0).AsEnumerable(Of Single).ToArray()
            Dim norf As New List(Of Byte)
            For Each data As Single In outputData
                Dim val As Integer = If(data > boxThresh, 255, 0)
                norf.Add(CByte(val))
            Next
            Dim fMapMat As New Mat(rows, cols, DepthType.Cv32F, 1)
            fMapMat.SetTo(outputData)
            Console.WriteLine(fMapMat)

            Dim norfMapMat As New Mat(rows, cols, DepthType.Cv8U, 1)
            norfMapMat.SetTo(norf.ToArray())
            Console.WriteLine(norfMapMat)

            Dim contours As New VectorOfVectorOfPoint()

            CvInvoke.FindContours(norfMapMat, contours, Nothing, RetrType.List, ChainApproxMethod.ChainApproxSimple)
            For i As Integer = 0 To contours.Size - 1
                Dim minEdgeSize As Single = 0
                Dim box As List(Of PointF) = GetMiniBox(contours(i), minEdgeSize)
                If minEdgeSize < minArea Then
                    Continue For
                End If
                Dim score As Double = GetScore(contours(i), fMapMat)
                If score < CDbl(boxScoreThresh) Then
                    Continue For
                End If
                Dim newBox As List(Of Point) = Unclip(box, unClipRatio)
                If newBox Is Nothing Then
                    Continue For
                End If

                Dim minBox As List(Of PointF) = GetMiniBox(newBox, minEdgeSize)
                If minEdgeSize < minArea + 2 Then
                    Continue For
                End If
                Dim finalPoints As New List(Of Point)
                For Each item As PointF In minBox
                    Dim x As Integer = CInt(item.X / s.ScaleWidth)
                    Dim ptx As Integer = Math.Min(Math.Max(x, 0), s.SrcWidth)

                    Dim y As Integer = CInt(item.Y / s.ScaleHeight)
                    Dim pty As Integer = Math.Min(Math.Max(y, 0), s.SrcHeight)
                    Dim dstPt As New Point(ptx, pty)
                    finalPoints.Add(dstPt)
                Next

                Dim textBox As New TextBox(finalPoints, CSng(score))
                rsBoxes.Add(textBox)
            Next
            rsBoxes.Reverse()
            Return rsBoxes
        End Function

        Private Shared Function GetMiniBox(contours As List(Of Point), ByRef minEdgeSize As Single) As List(Of PointF)
            Dim vop As New VectorOfPoint()
            vop.Push(contours.ToArray())
            Return GetMiniBox(vop, minEdgeSize)
        End Function

        Private Shared Function GetMiniBox(contours As VectorOfPoint, ByRef minEdgeSize As Single) As List(Of PointF)
            Dim box As New List(Of PointF)
            Dim rrect As RotatedRect = CvInvoke.MinAreaRect(contours)
            Dim points As PointF() = CvInvoke.BoxPoints(rrect)
            minEdgeSize = Math.Min(rrect.Size.Width, rrect.Size.Height)

            Dim thePoints As New List(Of PointF)(points)
            thePoints.Sort(AddressOf CompareByX)

            Dim index_1 As Integer = 0, index_2 As Integer = 1, index_3 As Integer = 2, index_4 As Integer = 3
            If thePoints(1).Y > thePoints(0).Y Then
                index_1 = 0
                index_4 = 1
            Else
                index_1 = 1
                index_4 = 0
            End If

            If thePoints(3).Y > thePoints(2).Y Then
                index_2 = 2
                index_3 = 3
            Else
                index_2 = 3
                index_3 = 2
            End If

            box.Add(thePoints(index_1))
            box.Add(thePoints(index_2))
            box.Add(thePoints(index_3))
            box.Add(thePoints(index_4))

            Return box
        End Function

        Public Shared Function CompareByX(left As PointF, right As PointF) As Integer
            If left = Nothing AndAlso right = Nothing Then
                Return 1
            End If

            If left = Nothing Then
                Return 0
            End If

            If right = Nothing Then
                Return 1
            End If

            If left.X > right.X Then
                Return 1
            End If

            If left.X = right.X Then
                Return 0
            End If

            Return -1
        End Function

        Private Shared Function GetScore(contours As VectorOfPoint, fMapMat As Mat) As Double
            Dim xmin As Short = 9999
            Dim xmax As Short = 0
            Dim ymin As Short = 9999
            Dim ymax As Short = 0

            Try
                For Each point As Point In contours.ToArray()
                    If point.X < xmin Then
                        'var xx = nd[point.X];
                        xmin = CShort(point.X)
                    End If

                    If point.X > xmax Then
                        xmax = CShort(point.X)
                    End If

                    If point.Y < ymin Then
                        ymin = CShort(point.Y)
                    End If

                    If point.Y > ymax Then
                        ymax = CShort(point.Y)
                    End If
                Next

                Dim roiWidth As Integer = xmax - xmin + 1
                Dim roiHeight As Integer = ymax - ymin + 1

                Dim bitmap As Image(Of Gray, Single) = fMapMat.ToImage(Of Gray, Single)
                Dim roiBitmap As New Image(Of Gray, Single)(roiWidth, roiHeight)
                Dim dataFloat As Single(,,) = bitmap.Data
                Dim data As Single(,,) = roiBitmap.Data

                For j As Integer = ymin To ymin + (roiHeight - 1)
                    For i As Integer = xmin To xmin + (roiWidth - 1)
                        Try
                            data(j - ymin, i - xmin, 0) = dataFloat(j, i, 0)
                        Catch ex2 As Exception
                            Console.WriteLine(ex2.Message)
                        End Try
                    Next
                Next

                Dim mask As Mat = Mat.Zeros(roiHeight, roiWidth, DepthType.Cv8U, 1)
                Dim pts As New List(Of Point)
                For Each point As Point In contours.ToArray()
                    pts.Add(New Point(point.X - xmin, point.Y - ymin))
                Next

                Using vp As New VectorOfPoint(pts.ToArray())
                    Using vvp As New VectorOfVectorOfPoint(vp)
                        CvInvoke.FillPoly(mask, vvp, New MCvScalar(1))
                    End Using
                End Using

                Return CvInvoke.Mean(roiBitmap, mask).V0
            Catch ex As Exception
                Console.WriteLine(ex.Message + ex.StackTrace)
            End Try

            Return 0
        End Function

        Private Shared Function Unclip(box As List(Of PointF), unclip_ratio As Single) As List(Of Point)
            Dim theCliperPts As New List(Of IntPoint)
            For Each pt As PointF In box
                Dim a1 As New IntPoint(CInt(pt.X), CInt(pt.Y))
                theCliperPts.Add(a1)
            Next

            Dim area As Single = Math.Abs(SignedPolygonArea(box.ToArray()))
            Dim length As Double = LengthOfPoints(box)
            Dim distance As Double = CDbl(area) * CDbl(unclip_ratio) / length

            Dim co As New ClipperOffset()
            co.AddPath(theCliperPts, JoinType.jtRound, EndType.etClosedPolygon)
            Dim solution As New List(Of List(Of IntPoint))
            co.Execute(solution, distance)
            If solution.Count = 0 Then
                Return Nothing
            End If

            Dim retPts As New List(Of Point)
            For Each ip As IntPoint In solution(0)
                retPts.Add(New Point(CInt(ip.X), CInt(ip.Y)))
            Next

            Return retPts
        End Function

        Private Shared Function SignedPolygonArea(Points As PointF()) As Single
            ' Add the first point to the end.
            Dim num_points As Integer = Points.Length
            Dim pts As PointF() = New PointF(num_points) {}
            Points.CopyTo(pts, 0)
            pts(num_points) = Points(0)

            ' Get the areas.
            Dim area As Single = 0
            For i As Integer = 0 To num_points - 1
                area += (pts(i + 1).X - pts(i).X) * (pts(i + 1).Y + pts(i).Y) / 2
            Next

            Return area
        End Function

        Private Shared Function LengthOfPoints(box As List(Of PointF)) As Double
            Dim length As Double = 0

            Dim pt As PointF = box(0)
            Dim x0 As Double = CDbl(pt.X)
            Dim y0 As Double = CDbl(pt.Y)
            Dim x1 As Double = 0, y1 As Double = 0, dx As Double = 0, dy As Double = 0
            box.Add(pt)

            Dim count As Integer = box.Count
            For idx As Integer = 1 To count - 1
                Dim pts As PointF = box(idx)
                x1 = CDbl(pts.X)
                y1 = CDbl(pts.Y)
                dx = x1 - x0
                dy = y1 - y0

                length += Math.Sqrt(dx * dx + dy * dy)

                x0 = x1
                y0 = y1
            Next

            box.RemoveAt(count - 1)
            Return length
        End Function
    End Class
End Namespace

File: dotnet_projects/OcrLiteOnnxCs/OcrLiteLib/DbNet.cs
using ClipperLib;
using Emgu.CV;
using Emgu.CV.CvEnum;
using Emgu.CV.Structure;
using Emgu.CV.Util;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace OcrLiteLib
{
    class DbNet
    {
        private readonly float[] MeanValues = { 0.485F * 255F, 0.456F * 255F, 0.406F * 255F };
        private readonly float[] NormValues = { 1.0F / 0.229F / 255.0F, 1.0F / 0.224F / 255.0F, 1.0F / 0.225F / 255.0F };

        private InferenceSession dbNet;

        public DbNet() { }

        ~DbNet()
        {
            dbNet.Dispose();
        }

        public void InitModel(string path, int numThread)
        {
            try
            {
                SessionOptions op = new SessionOptions();
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED;
                op.InterOpNumThreads = numThread;
                op.IntraOpNumThreads = numThread;
                dbNet = new InferenceSession(path, op);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
                throw ex;
            }
        }

        public List<TextBox> GetTextBoxes(Mat src, ScaleParam scale, float boxScoreThresh, float boxThresh, float unClipRatio)
        {
            Mat srcResize = new Mat();
            CvInvoke.Resize(src, srcResize, new Size(scale.DstWidth, scale.DstHeight));
            Tensor<float> inputTensors = OcrUtils.SubstractMeanNormalize(srcResize, MeanValues, NormValues);
            var inputs = new List<NamedOnnxValue>
            {
                NamedOnnxValue.CreateFromTensor("input0", inputTensors)
            };
            try
            {
                using (IDisposableReadOnlyCollection<DisposableNamedOnnxValue> results = dbNet.Run(inputs))
                {
                    var resultsArray = results.ToArray();
                    Console.WriteLine(resultsArray);
                    var textBoxes = GetTextBoxes(resultsArray, srcResize.Rows, srcResize.Cols, scale, boxScoreThresh, boxThresh, unClipRatio);
                    return textBoxes;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
            }
            return null;
        }

        private static List<TextBox> GetTextBoxes(DisposableNamedOnnxValue[] outputTensor, int rows, int cols, ScaleParam s, float boxScoreThresh, float boxThresh, float unClipRatio)
        {
            float minArea = 3.0f;
            List<TextBox> rsBoxes = new List<TextBox>();

            float[] outputData = outputTensor[0].AsEnumerable<float>().ToArray();
            List<byte> norf = new List<byte>();
            foreach (float data in outputData)
            {
                int val = data > boxThresh ? 255 : 0;
                norf.Add((byte)val);
            }
            Mat fMapMat = new Mat(rows, cols, DepthType.Cv32F, 1);
            fMapMat.SetTo(outputData);
            Console.WriteLine(fMapMat);

            Mat norfMapMat = new Mat(rows, cols, DepthType.Cv8U, 1);
            norfMapMat.SetTo(norf.ToArray());
            Console.WriteLine(norfMapMat);

            VectorOfVectorOfPoint contours = new VectorOfVectorOfPoint();

            CvInvoke.FindContours(norfMapMat, contours, null, RetrType.List, ChainApproxMethod.ChainApproxSimple);
            for (int i = 0; i < contours.Size; i++)
            {
                float minEdgeSize = 0;
                List<PointF> box = GetMiniBox(contours[i], out minEdgeSize);
                if (minEdgeSize < minArea)
                {
                    continue;
                }
                double score = GetScore(contours[i], fMapMat);
                if (score < boxScoreThresh)
                {
                    continue;
                }
                List<Point> newBox = Unclip(box, unClipRatio);
                if (newBox == null)
                {
                    continue;
                }

                List<PointF> minBox = GetMiniBox(newBox, out minEdgeSize);
                if (minEdgeSize < minArea + 2)
                {
                    continue;
                }
                List<Point> finalPoints = new List<Point>();
                foreach (var item in minBox)
                {
                    int x = (int)(item.X / s.ScaleWidth);
                    int ptx = Math.Min(Math.Max(x, 0), s.SrcWidth);

                    int y = (int)(item.Y / s.ScaleHeight);
                    int pty = Math.Min(Math.Max(y, 0), s.SrcHeight);
                    Point dstPt = new Point(ptx, pty);
                    finalPoints.Add(dstPt);
                }

                TextBox textBox = new TextBox();
                textBox.Score = (float)score;
                textBox.Points = finalPoints;
                rsBoxes.Add(textBox);
            }
            rsBoxes.Reverse();
            return rsBoxes;
        }

        private static List<PointF> GetMiniBox(List<Point> contours, out float minEdgeSize)
        {
            VectorOfPoint vop = new VectorOfPoint();
            vop.Push(contours.ToArray<Point>());
            return GetMiniBox(vop, out minEdgeSize);
        }

        private static List<PointF> GetMiniBox(VectorOfPoint contours, out float minEdgeSize)
        {
            List<PointF> box = new List<PointF>();
            RotatedRect rrect = CvInvoke.MinAreaRect(contours);
            PointF[] points = CvInvoke.BoxPoints(rrect);
            minEdgeSize = Math.Min(rrect.Size.Width, rrect.Size.Height);

            List<PointF> thePoints = new List<PointF>(points);
            thePoints.Sort(CompareByX);

            int index_1 = 0, index_2 = 1, index_3 = 2, index_4 = 3;
            if (thePoints[1].Y > thePoints[0].Y)
            {
                index_1 = 0;
                index_4 = 1;
            }
            else
            {
                index_1 = 1;
                index_4 = 0;
            }

            if (thePoints[3].Y > thePoints[2].Y)
            {
                index_2 = 2;
                index_3 = 3;
            }
            else
            {
                index_2 = 3;
                index_3 = 2;
            }

            box.Add(thePoints[index_1]);
            box.Add(thePoints[index_2]);
            box.Add(thePoints[index_3]);
            box.Add(thePoints[index_4]);

            return box;
        }

        public static int CompareByX(PointF left, PointF right)
        {
            if (left == null && right == null)
            {
                return 1;
            }

            if (left == null)
            {
                return 0;
            }

            if (right == null)
            {
                return 1;
            }

            if (left.X > right.X)
            {
                return 1;
            }

            if (left.X == right.X)
            {
                return 0;
            }

            return -1;
        }

        private static double GetScore(VectorOfPoint contours, Mat fMapMat)
        {
            short xmin = 9999;
            short xmax = 0;
            short ymin = 9999;
            short ymax = 0;

            try
            {
                foreach (Point point in contours.ToArray())
                {
                    if (point.X < xmin)
                    {
                        //var xx = nd[point.X];
                        xmin = (short)point.X;
                    }

                    if (point.X > xmax)
                    {
                        xmax = (short)point.X;
                    }

                    if (point.Y < ymin)
                    {
                        ymin = (short)point.Y;
                    }

                    if (point.Y > ymax)
                    {
                        ymax = (short)point.Y;
                    }
                }

                int roiWidth = xmax - xmin + 1;
                int roiHeight = ymax - ymin + 1;

                Image<Gray, float> bitmap = fMapMat.ToImage<Gray, float>();
                Image<Gray, float> roiBitmap = new Image<Gray, float>(roiWidth, roiHeight);
                float[,,] dataFloat = bitmap.Data;
                float[,,] data = roiBitmap.Data;

                for (int j = ymin; j < ymin + roiHeight; j++)
                {
                    for (int i = xmin; i < xmin + roiWidth; i++)
                    {
                        try
                        {
                            data[j - ymin, i - xmin, 0] = dataFloat[j, i, 0];
                        }
                        catch (Exception ex2)
                        {
                            Console.WriteLine(ex2.Message);
                        }
                    }
                }

                Mat mask = Mat.Zeros(roiHeight, roiWidth, DepthType.Cv8U, 1);
                List<Point> pts = new List<Point>();
                foreach (Point point in contours.ToArray())
                {
                    pts.Add(new Point(point.X - xmin, point.Y - ymin));
                }

                using (VectorOfPoint vp = new VectorOfPoint(pts.ToArray<Point>()))
                using (VectorOfVectorOfPoint vvp = new VectorOfVectorOfPoint(vp))
                {
                    CvInvoke.FillPoly(mask, vvp, new MCvScalar(1));
                }

                return CvInvoke.Mean(roiBitmap, mask).V0;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
            }

            return 0;
        }

        private static List<Point> Unclip(List<PointF> box, float unclip_ratio)
        {
            List<IntPoint> theCliperPts = new List<IntPoint>();
            foreach (PointF pt in box)
            {
                IntPoint a1 = new IntPoint((int)pt.X, (int)pt.Y);
                theCliperPts.Add(a1);
            }

            float area = Math.Abs(SignedPolygonArea(box.ToArray<PointF>()));
            double length = LengthOfPoints(box);
            double distance = area * unclip_ratio / length;

            ClipperOffset co = new ClipperOffset();
            co.AddPath(theCliperPts, JoinType.jtRound, EndType.etClosedPolygon);
            List<List<IntPoint>> solution = new List<List<IntPoint>>();
            co.Execute(ref solution, distance);
            if (solution.Count == 0)
            {
                return null;
            }

            List<Point> retPts = new List<Point>();
            foreach (IntPoint ip in solution[0])
            {
                retPts.Add(new Point((int)ip.X, (int)ip.Y));
            }

            return retPts;
        }

        private static float SignedPolygonArea(PointF[] Points)
        {
            // Add the first point to the end.
            int num_points = Points.Length;
            PointF[] pts = new PointF[num_points + 1];
            Points.CopyTo(pts, 0);
            pts[num_points] = Points[0];

            // Get the areas.
            float area = 0;
            for (int i = 0; i < num_points; i++)
            {
                area +=
                    (pts[i + 1].X - pts[i].X) *
                    (pts[i + 1].Y + pts[i].Y) / 2;
            }

            return area;
        }

        private static double LengthOfPoints(List<PointF> box)
        {
            double length = 0;

            PointF pt = box[0];
            double x0 = pt.X;
            double y0 = pt.Y;
            double x1 = 0, y1 = 0, dx = 0, dy = 0;
            box.Add(pt);

            int count = box.Count;
            for (int idx = 1; idx < count; idx++)
            {
                PointF pts = box[idx];
                x1 = pts.X;
                y1 = pts.Y;
                dx = x1 - x0;
                dy = y1 - y0;

                length += Math.Sqrt(dx * dx + dy * dy);

                x0 = x1;
                y0 = y1;
            }

            box.RemoveAt(count - 1);
            return length;
        }

    }
}

